src/Microsoft.DotNet.Interactive.Tests/Parsing/SubmissionParserTests.cs

```diff
@@ -365,6 +366,15 @@ private static SubmissionParser CreateSubmissionParser(string defaultLanguage =
         return compositeKernel.SubmissionParser;
     }
 
+    [Fact]
+    public async Task DiagnosticsProduced_events_always_point_back_to_the_original_command()
```

Comment:
This is intended to be generally true, so not specific to `DiagnosticsProduced`. The internal command (`AnonymousKernelCommand` in this case, but it could as easily be one of the public command types) is an implementation detail that the external caller shouldn't have to be aware of.

It might be useful to add another similar test for the split between subkernels because the same should be true for, for example:

```csharp
#!csharp
123
#!fsharp
456
```

Reply:
So looks like this invariant is not satisfied for all events in the above case. For example, `CodeSubmissionReceived`, `CompleteCodeSubmissionReceived`, `ReturnValueProduced` events for the repro code you shared above point back to the per-language `SubmitCode` commands created internally for C# and F# respectively. Only the final `CommandSucceded` event points back to the original `SubmitCode` command.

If I change the submission to the following then interestingly, the `DiagnosticsProduced` event for the erroneous line points to the F#-specific `SubmitCode` command with code `some error`. However, the line spans in the `DiagnosticsProduced.Diagnostics` reflect the spans in the original submission.
```
#!csharp
123
 #!fsharp
some error
```

This may be *reasonable* considering that the inner commands reflect the expected command splitting behavior - however, it also seems like internal implementation details of the callee are bleeding through to the caller.

If this is unexpected, I can log a separate bug to a) fix the behavior and b) add tests to ensure that inner commands are never exposed to the caller. Thoughts?

Reply:
@jonsequitur Btw the current PR should be ready to merge. The above issue needs more investigation / follow-up and I'd prefer not to hold up the current PR for that. Please approve if you agree :) Thanks!